### PR TITLE
Introduce higher-level API for py client

### PR DESCRIPTION
**Pull request by antisaling**
## Description

Introduce the first iteration of a higher-level Python API as proposed on #129.

Related: #129
Depends on: #209

## How Has This Been Tested?

New tests were introduced on `tests/test_client.py`.

## Merge criteria:

- [x] The commits and have meaningful messages; the author will squash them [after approval](https://github.com/opendatahub-io/opendatahub-community/blob/main/contributor-cheatsheet.md#:~:text=Usually%20this%20is%20done%20in%20last%20phase%20of%20a%20PR%20revision) or will ask to merge with squash.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: clients/python/src/model_registry/_client.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+    def _register_model_artifact(
+        self, mv: ModelVersion, uri: str, /, **kwargs
+    ) -> ModelArtifact:
+        assert mv.id is not None, "Model version must have an ID"
+        ma = ModelArtifact(mv.model_name, uri, **kwargs)
+        self._api.upsert_model_artifact(ma, mv.id)
+        return ma
+
+    def register_model(
```
Comment (tarilabs) [inline clients/python/src/model_registry/_client.py:65]: should we rename this ~"register model artifact" or something else to better indicate this is performing the whole chain of inserting (or reusing) RegisteredModel, ModelVersion and ModelArtifact? 🤔

**Review by tarilabs — DISMISSED** [review]
minor comment which I believe is better to address before merging, but I don't want to hold merging solely based on that.

Thank you so much this will be extremely helpful

File: clients/python/src/model_registry/_client.py (hunk 1 above)
Comment (lampajr) [reply clients/python/src/model_registry/_client.py:65]: In this regard IMO `register_model` is good enough, I found the same pattern on [mlflow api](https://mlflow.org/docs/latest/model-registry.html#api-workflow):
- `mlflow.register_model` high level api to register the full model (model, version and artifact)
- `client.create_registered_model` raw api to create empty model - the same as our `upsert_registered_model`
- `client.create_model_version` raw api to create a new model version - the same as our `upsert_model_version`

wdyt?

File: clients/python/src/model_registry/__init__.py [hunk 2]
```diff
@@ -2,11 +2,8 @@
 
 __version__ = "0.1.0"
 
-from .client import ModelRegistry
-from .types import ListOptions, OrderByField
+from ._client import ModelRegistry
```
Comment (lampajr) [inline clients/python/src/model_registry/__init__.py:5]: So if I correctly understood, with this we are going to "export" just the high level API and not the low level one (i.e., `ModelRegistryAPIClient`) - this way users cannot directly use the underlying API, right?

If I am right, is this what we want? I am thinking how user could update for instance a model artifact metadata if we are not exposing the underlying api.

If I am not right, forget what I said :)

File: clients/python/src/model_registry/_client.py (hunk 1 above)
Comment (tarilabs) [reply clients/python/src/model_registry/_client.py:65]: yes, thank you for making me notice that, lgtm 👍

File: clients/python/src/model_registry/__init__.py (hunk 2 above)
Comment (antisaling) [reply clients/python/src/model_registry/__init__.py:5]: Yes, by default we only export the higher-level API, but the user can also do `from model_registry.core import ModelRegistryAPIClient as MRAPI` and use the lower-level bindings the same way they would in go.

Edit: clarification

> If I am right, is this what we want? I am thinking how user could update for instance a model artifact metadata if we are not exposing the underlying api.

I'm actually planning on another more powerful version of this higher-level API, but if it's really important that this is exposed by default, we can think of workarounds :)

Comment (lampajr) [reply clients/python/src/model_registry/__init__.py:5]: Well actually if the user can import the lower level api that is fine for me, make sense to export by default just the higer level one.

But we should guarantee a way to make use of the lower level if the user want to use it, at least for now I think

Comment (tarilabs) [reply clients/python/src/model_registry/__init__.py:5]: > But we should guarantee a way to make use of the lower level if the user want to use it, at least for now I think

+1

Thanks for the clarifications Isabella

**Review by tarilabs — APPROVED** [review]

**Review by lampajr — APPROVED** [review]

**Pull request merged**